tools/plagiarism-checker/src/index.ts

```diff
@@ -0,0 +1,93 @@
+import { Hono } from "hono"
+
+interface SearchResult {
+  title?: string | null
+  link?: string | null
+}
+
+interface SearchResponse {
+  items: Array<SearchResult>
+}
+
+interface PlagiarismResults {
+  sentence: string
+  matches: Array<SearchResult>
+}
+
+type Env = {
+  GOOGLE_API_KEY: string
+  GOOGLE_SEARCH_ENGINE_CX: string
+}
+
+const app = new Hono<{ Bindings: Env }>()
+
+app.use("*", async (c, next) => {
+  const apiKey = c.env.GOOGLE_API_KEY
+  if (!apiKey) {
+    return c.text("API key not found", 400)
+  }
+
+  const searchEngine = c.env.GOOGLE_SEARCH_ENGINE_CX
+  if (!searchEngine) {
+    return c.text("Search engine not specified", 400)
+  }
+
+  return next()
+})
+
+app.post("/", async (c) => {
+  const { text } = await c.req.json()
+
+  if (typeof text !== "string") {
+    return c.json({ error: "Invalid JSON format" }, 400)
+  }
+
+  const sentences = text
+    .split(".")
+    // if it's less than 10 characters - we don't treat it as a sentence and skip it
+    .filter((sentence) => sentence.trim().length > 10)
+
+  const plagiarismResults: PlagiarismResults[] = []
+
+  await Promise.all(
+    sentences.map(async (sentence) => {
+      const query = encodeURIComponent(sentence.trim())
+      const url = `https://www.googleapis.com/customsearch/v1?exactTerms=${query}&key=${c.env.GOOGLE_API_KEY}&cx=${c.env.GOOGLE_SEARCH_ENGINE_CX}&num=1`
+      const response = await fetch(url)
+      const searchResults: SearchResponse = await response.json()
+
+      if (!searchResults.items) {
+        return
+      }
+      plagiarismResults.push({
+        sentence,
+        matches: searchResults.items.map((item: SearchResult) => ({
+          title: item.title,
+          link: item.link
+        }))
+      })
+    })
+  )
+
+  let plagiarisedSentencesNum = 0
+
+  if (plagiarismResults) {
+    plagiarisedSentencesNum = plagiarismResults.reduce((total, current) => {
+      if (current?.matches?.length && current?.matches?.length > 0) {
+        total++
+      }
+      return total
+    }, 0)
+  }
+
+  const plagiarismPercent = plagiarisedSentencesNum
+    ? (sentences.length / plagiarisedSentencesNum) * 100
+    : 0
```

Comment:
Please check the formula.

Reply:
> Please check the formula.

It might look a bit confusing, but I think it's correct. 

We don't want to divide by 0, so on the first line we are checking if plagiarisedSentencesNum is falsy (0). 

If it's not, we are dividing amount of sentences by amount of plagiarized sentences and multiplying by 100 to turn into % form.

If it's falsy - the percent is set to 0.

Also, there might be an edge-case where sentences length is zero, but I handled that earlier in the code in my latest commits

Reply:
Wait, I got it... Has to be plagiarisedSentencesNum / sentences.length instead.
Fixed it. My bad :)